File: leetcode/solutions/0109.ConvertSortedListToBinarySearchTree/sortedListToBST_test.ts
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
import sortedListToBST from "./sortedListToBST.ts";
import {
  SinglyLinkedListNode,
  CreateListNode,
} from "../../../datastructure/SinglyLinkedListNode.ts";
import {
  BinaryTreeNode,
  createBinaryTreeNode,
} from "../../../datastructure/BinaryTreeNode.ts";

Deno.test("0109. Convert Sorted List to Binary Search Tree", () => {
  type test = [
    SinglyLinkedListNode<number> | null,
    BinaryTreeNode<number> | null
  ];
  const tests: test[] = [
    [CreateListNode([0]), createBinaryTreeNode([0])],
    [CreateListNode([-1, 0, 1]), createBinaryTreeNode([0, -1, 1])],
    [
      CreateListNode([-10, -3, 0, 5, 9]),
      createBinaryTreeNode([0, -10, 5, null, -3, null, 9]),
    ],
  ];

  for (const [arg, want] of tests) {
    assertEquals(sortedListToBST(arg), want);
  }
});

File: leetcode/solutions/0109.ConvertSortedListToBinarySearchTree/sortedListToBST_test.go
package solution

import (
	"testing"

	"github.com/google/go-cmp/cmp"
	ds "github.com/yukisato/leetcode-util/datastructure"
)

func Test_0109_ConvertSortedListToBinarySearchTree(t *testing.T) {
	t.Parallel()
	tests := []struct {
		in  *ListNode
		out *TreeNode
	}{
		{
			ds.CreateListNode([]int{0}),
			ds.CreateTreeNodeFromString([]string{"0"}),
		},
		{
			ds.CreateListNode([]int{-1, 0, 1}),
			ds.CreateTreeNodeFromString([]string{"0", "-1", "1"}),
		},
		{
			ds.CreateListNode([]int{-10, -3, 0, 5, 9}),
			ds.CreateTreeNodeFromString([]string{"0", "-3", "9", "-10", "null", "5"}),
		},
	}

	for _, tt := range tests {
		tt := tt

		t.Run("Test", func(t *testing.T) {
			t.Parallel()

			got := sortedListToBST(tt.in)

			if !cmp.Equal(got, tt.out) {
				t.Errorf("got %#v want %#v", got, tt.out)
			}
		})
	}
}
